### DISCUSSION TITLE
Is it legal to use refine on an object field ?

### DISCUSSION THREAD

**Discussion by nhedger** [colinhacks/zod#2335] [Q&A]
I am trying to parse an object that has an `id` field that must be a `string` in `uuid` format. I want to use refine to do some additional validation on the value, but I end up in `refine` even if `id` does not contain a valid `uuid` string.

```ts
const schema = z.object({
  id: z.string().uuid().refine(() => {
    throw new Error("should not be called");
  }),
})

const result = schema.safeParse({id: "not a uuid"});
```

I tried to do something similar without the UUID validation and that actually works. In the following case I do NOT end up in refine, as expected.

```ts
const schema = z.object({
  id: z.string().refine(() => {
    throw new Error("should not be called");
  }),
})

const result = schema.safeParse({id: 0});
```

I can provide a failing test for my use case if needed.

Is this behavior expected ?

**Comment by JacobWeisenburger** [comment]
Related:
https://github.com/colinhacks/zod/issues/2243
https://github.com/colinhacks/zod/issues/2192
https://github.com/colinhacks/zod/issues/2113

**Reply by nhedger** [reply]
Thanks for linking. I actually saw these issues but wasn't sure if my case was linked.
